File: backend/api/repositories/review_repository.py
"""Review repository for database operations"""
from typing import List, Dict
from .base import BaseRepository


class ReviewRepository(BaseRepository):
    """Repository for review operations"""
    
    def __init__(self, db):
        super().__init__(db, "reviews")
    
    async def find_by_product(self, product_id: str, status: str = "approved",
                             limit: int = 20, skip: int = 0, sort_by: str = "recent") -> List[Dict]:
        """Find reviews for a product"""
        query = {"product_id": product_id, "status": status}
        
        sort_options = {
            "recent": [("created_at", -1)],
            "helpful": [("helpful_count", -1)],
            "rating_high": [("rating", -1)],
            "rating_low": [("rating", 1)]
        }
        
        sort = sort_options.get(sort_by, [("created_at", -1)])
        return await self.find_many(query, limit=limit, skip=skip, sort=sort)
    
    async def find_by_user_and_product(self, user_id: str, product_id: str) -> Dict:
        """Check if user reviewed a product"""
        return await self.find_one({"user_id": user_id, "product_id": product_id})
    
    async def get_product_rating_stats(self, product_id: str) -> Dict:
        """Get rating statistics for a product"""
        pipeline = [
            {"$match": {"product_id": product_id, "status": "approved"}},
            {
                "$group": {
                    "_id": None,
                    "avg_rating": {"$avg": "$rating"},
                    "total_reviews": {"$sum": 1}
                }
            }
        ]
        result = await self.aggregate(pipeline)
        return result[0] if result else {"avg_rating": 0, "total_reviews": 0}

File: nextjs-app/src/lib/repositories/review.repository.ts
import { BaseRepository } from './base.repository';

interface Review {
  id?: string;
  product_id: string;
  user_id: string;
  rating: number;
  comment: string;
  helpful_count?: number;
  created_at?: Date;
  updated_at?: Date;
}

export class ReviewRepository extends BaseRepository<Review> {
  constructor() {
    super('reviews');
  }

  async findByProductId(productId: string): Promise<Review[]> {
    return this.findAll(
      { product_id: productId },
      { sort: { created_at: -1 } }
    );
  }

  async findByUserAndProduct(userId: string, productId: string): Promise<Review | null> {
    return this.findOne({ user_id: userId, product_id: productId });
  }

  async getAverageRating(productId: string): Promise<number> {
    const db = await (await import('@/lib/db')).getDatabase();
    const result = await db
      .collection(this.collectionName)
      .aggregate([
        { $match: { product_id: productId } },
        { $group: { _id: null, avg: { $avg: '$rating' } } },
      ])
      .toArray();

    return result[0]?.avg || 0;
  }
}
